core/finality-grandpa/src/communication/mod.rs

```diff
@@ -164,18 +196,45 @@ impl Stream for NetworkStream {
 	}
 }
 
+/// The result of processing a commit.
+pub(crate) enum CommitProcessingOutcome {
+	Good,
+	Bad,
+}
+
 /// Bridge between the underlying network service, gossiping consensus messages and Grandpa
 pub(crate) struct NetworkBridge<B: BlockT, N: Network<B>> {
 	service: N,
 	validator: Arc<GossipValidator<B>>,
+	neighbor_sender: periodic::NeighborPacketSender<B>,
 }
 
 impl<B: BlockT, N: Network<B>> NetworkBridge<B, N> {
-	/// Create a new NetworkBridge to the given NetworkService
-	pub(crate) fn new(service: N, config: crate::Config) -> Self {
-		let validator = Arc::new(GossipValidator::new(config));
+	/// Create a new NetworkBridge to the given NetworkService. Returns the service
+	/// handle and a future that must be polled to completion to finish startup.
+	pub(crate) fn new(service: N, config: crate::Config) -> (
+		Self,
+		impl futures::Future<Item = (), Error = ()> + Send + 'static,
+	) {
+
+		let (validator, report_stream) = GossipValidator::new(config);
+		let validator = Arc::new(validator);
 		service.register_validator(validator.clone());
-		NetworkBridge { service, validator: validator }
+
+		let (rebroadcast_job, neighbor_sender) = periodic::neighbor_packet_worker(service.clone());
+		let reporting_job = report_stream.consume(service.clone());
+
+		let bridge = NetworkBridge { service, validator, neighbor_sender };
+
+		let startup_work = futures::future::lazy(move || {
+			// lazily spawn these jobs onto their own tasks. the lazy future has access
+			// to tokio globals, which aren't available outside.
+			tokio::spawn(rebroadcast_job);
```

Comment:
These should be exit scoped, I think that's what's making the connectivity test fail.